### DISCUSSION TITLE
Proposal: propsToModels()

### DISCUSSION THREAD

**Discussion by Vasile-Peste** [vuejs/rfcs#725] [General]
The idea is that `defineProps()` should actually return a mutable copy of the actual props, with listeners on them automatically emitting when the props copies are modified or when the original props are modified.

The API would be something like `propsToModels()`

```javascript
/*
 * Used to create mutable refs from props, with reactive two way binding
 */
export function propsToModels (props, emit) {
    const refs = {};
    let propsCache = cloneDeep(toRaw(props));

    for (const propName of Object.keys(props)) {
        const propMirror = ref(reactive(cloneDeep(props[propName])));

        watch(propMirror, (value) => {
            emit(`update:${propName}`, cloneDeep(value));
        }, {
            deep: true,
        });

        refs[propName] = propMirror;
    }

    watch(props, (value) => {
        for (const propName of Object.keys(props)) {
            const cachedValue = JSON.stringify(propsCache[propName]);

            if (cachedValue !== JSON.stringify(props[propName])) {
                const newValue = cloneDeep(toRaw(value[propName]));

                refs[propName].value = newValue;
                propsCache[propName] = newValue;
            }
        }
    }, {
        deep: true,
    });

    return refs;
}
```

This is a rough implementation and can be improved a lot

This solves the issue and need of having to create a local copy of the prop and listening to its mutations in order to update the real prop. It's particularly useful when dealing with nested v-model props and when props are objects.

An example

```javascript
const props = defineProps({
    workspace: {
        type: Object,
        required: true,
    },
});
const {
    workspace,
} = propsToModels(props, emit);

// Now we can do <MyComponent v-model="workspace"/>
// and even
// <MyComponent v-model="workspace.property.property.property"/>
```

Now `workspace` can be deeply used with v-model by child components and so on. Mutating `workspace` will trigger an emit which will update the real `workspace` prop. If the `workspace` prop is mutated by parent or other component, the local copy will be updated too and remain mutable.

**Comment by posva** [comment]
It seems that you are looking for [`defineModel()`](https://vuejs.org/api/sfc-script-setup.html#definemodel)

**Comment by Vasile-Peste** [comment]
`defineModel()` doesn't work when using v-model on nested object properties
See also https://github.com/orgs/vuejs/discussions/10538

```javascript
<script setup>
import GrandChild from "./GrandChild.vue";

const model = defineModel();
</script>

<template>
  <div>
    <h3>Component</h3>
    a: <GrandChild v-model="model.a" /><br />
    b: <GrandChild v-model="model.b" />
  </div>
</template>
```

This doesn't trigger the emit and actually mutates the prop

**Reply by posva** [reply]
It seems like you want multiple `defineModel()`. Passing a property of an object will always mutate the object

**Reply by Vasile-Peste** [reply]
How would that work? We can't have multiple `defineModel()` if model is a single "big" object

**Reply by nickmessing** [reply]
@Vasile-Peste Destructuring an object into reactive computeds should not be part of the framework, and any tool that can do it will have edge cases, you can easily solve this in user space if you are okay with the limitations like in [this example](https://play.vuejs.org/#eNqtVc1um0AQfpURFxPFwUnbEyVWmzaHtGpSJVYvIQcMY4cY76LdxXaEePfO7gLG1K5TqT4Y2Pnm75ufLZ3Pee6tCnR8J5CxSHMFElWRQxax+WXoKBk645Cly5wLBSXI5yjL+PoeZ1DBTPAlDEh78LGFXKXzL5xeGTJVA7xR91B76+BLWEcqfu5bC1nMmVTApy8YK7jsOHbLkAEs8PXCh4HkSwSpRMrmg2F9/s6Hi/PzkFUn2o4x71o7Q3B5lvyKsgKHwHBt3k7gcgzGpnbJM/QyPncHBBwMoYGTqZ6c1EneGrHugpElkSijD4XLPIsU0hdAkAscl2WTUVUFI31iRDukrc6WPMGMyLfQ0IERwYJRx5wzpMpQOLN07r1Izqh8JoPQiclOmqG4y1VK4YaOb3PTMsPgN3OmBFHQnMfPGC/2nL/IjT4LnZ8CJYoVhk4rU5GYI8WmxdcPt7ih91ZICRQZof8ivEeistAxWthVwRIKu4Mz0d6YLqHqTuT1RiGTTVI6UI2sDD50qG80hYdS34b73vtg9KhgxKKlmLoKBbIYJ/w7vmLSfktPSeK27VXNbqEw2WlXXesaoV5zJFh/PDTCiO6MuwkngAwmQMQgSyThYy6SwPbxEAq2YHzNxmPqe5PBI0UFKdPNzWcwefK1h2Cij5+oGyiXkOHGRDArWKzTr/vscGaBNnw0hJCNXQ3sWasDIPGJ30uqM0sKqG+KTI9vWUH0hxNto81V5wAw4wJcq0vJ6px7nr2Vndm6stbBI2GfyEtTHrsi9I+K7oNrRnyvIaNZtxzo1ee3I31Qp6mK/XmetxfVGqXyaSdbw42AymYeZreYGup8VCFYwxtR1iPXdu2DIfA6SRUXb9jelk+zVSj2BGcpwx/6q67E2D2yurJoipl5BaA2uvDta5AyoruzsMyz3lckHjV6/d11WyynKP5D/MwY+tf46YIoS2OuqmsQTAulaGg+xVkaL5pMTk/J/Q2LBUYSg5HF1Gb2a5ydkcZX3KNxmIz+5fi2q7hLYXvP9nnt3LPdlmnx/T6yN2+77Y6tkHbHeaNj0O6dvq+QnTu93g7by9wWmTrf1Lmx0tHQ/8ZMR0n/N2cVOTsWoGuiOtJHSbpqyr/D53YC2lDaKSDsTq12sDbEzsRYD7stUv0GxINS4w==).

**Reply by Vasile-Peste** [reply]
Hi, the whole point is that the object comes from a prop. I don't see how this solution addresses the problem, can you elaborate it a bit further? Thanks!

**Reply by nickmessing** [reply]
If the object comes from a prop - you should not edit it, `prop` is something that child component should only read, like an argument to a function. If it's a `model` - my link has a tool that destructures into writable computeds (with the limitation that you need to have all keys and can't add/delete them)

**Reply by nickmessing** [reply]
In your example it comes from a model, just like my example in the link

**Reply by Vasile-Peste** [reply]
Try to pass object.key1 as v-model, in your example, the emit will not be triggered. This is the "problem" of defineModel() API (see also the linked discussion)

**Reply by nickmessing** [reply]
`defineModel` creates a writable reference with a function on `set` and another one on `get`, it doesn't make it "deep", and making it deep is expensive, in my example there's an utility that lets you "destructure reactively" in `key1` from `v-model` called `objectReferenceToKeyedReferences`, that's an example of how you can do that, but it's not "deep" either, it's only getting first level of it.

**Reply by nickmessing** [reply]
In my example link you can check `BigComponent.vue` for a usage example

**Reply by Vasile-Peste** [reply]
Thank you, however what we need is having it deep, as you mentioned. And I totally agree it's expensive, however on large scale applications with dozens of nested components, this feature would be necessary.

**Reply by nickmessing** [reply]
You can implement that in userspace, using proxy.
It's not a widespread usecase with way too big of a potential performance impact and to be honest it feels like an anti-pattern to me.
If I ever need something like this I'd rather make a utility extracting a single key like:
```vue
<script setup lang="ts">
const model = defineModel<MyObject>({ required: true })
const myField = useReactiveKey(model, 'key')
</script>
<template>
  <Something v-model="myField" />
</template>
```
since this is both explicit, easy to read/maintain/test and no significant impact on performance.

**Reply by jaypeerj** [reply]
> You can implement that in userspace, using proxy. It's not a widespread usecase with way too big of a potential performance impact and to be honest it feels like an anti-pattern to me. If I ever need something like this I'd rather make a utility extracting a single key

good suggestion, deep computed getters/setters are pretty good for this too. it can depend on what is known at runtime.  if you have a lot of keys you're obviously going to end up with a lot of extra code (whether that's a series of variable definitions or some kind of indexed collection/Object/etc.)

**Reply by alex-springer-sr** [reply]
> Destructuring an object into reactive computeds should not be part of the framework, and any tool that can do it will have edge cases, you can easily solve this in user space if you are okay with the limitations like in this example
@nickmessing could you elaborate on the edge cases and limitations of the example you provided?
I can only think of:
* not handling deep mutations
* all properties need to be defined from the start
which i am perfectly fine with for my use cases.

**Comment by edison1105** [comment]
It seems that you need https://vueuse.org/core/useVModels/

**Reply by nickmessing** [reply]
That makes a vModel for each prop, author wants vModels for each key of each prop, and then recursively deeper

**Reply by jacekkarczmarczyk** [reply]
There's https://vueuse.org/core/useVModel/ with `deep` option

**Reply by Vasile-Peste** [reply]
Yes! It would be great if Vue would support a `defineModel()` with `deep` option too

**Reply by jaypeerj** [reply]
buried in the `UseVModelOptions` `d.cts` is a bit more info telling you what you'll need

the below works as desired with props:

```
// note: passive needs to be true in my experience, be mindful that this uses a watcher under the hood
const d_foo = useVModel(
    props, 'foo', emit, { clone: cloneDeep,  passive: true,  deep: true })
```

here I'm using lodash's `cloneDeep` as my cloning algorithm. adjust as you see fit.

**Comment by michaelcozzolino** [comment]
i made a similar proposal, with a different api that leverages proxies and directives to solve the same thing
https://github.com/vuejs/rfcs/discussions/751
